topics/threaded-io.md

```diff
@@ -0,0 +1,119 @@
+
+## Configuration
+We have two options to control how I/O threads perform.
+```
+io-threads [thread-num]
+```
+The `io-threads` option control how many threads you want to use on reading/writing client connections.
+- 1 or not configured: Use single thread like what we have in the older version.
+- n (n >= 2): Use n threads(1 main thread & n-1 I/O threads) to handle network reads/writes.
+
+```
+io-threads-do-reads [yes|no]
+```
+By default, Redis uses I/O threads on the write system call only. If you want to use it on the read  system call, you have to enable this option as well. Usually, threaded reads don't help much.
+
+## High-Level Architecture
+The main idea of this feature is to postpone the read and write action and accumulate those events to the next event loop cycle. So that we have multiple events waiting to be handled and use I/O threads to deal with them.
+
+For example, when Redis has done executing a command from client `A` and got the reply content, instead of writing it to the client connection synchronously, Redis save it to a **pending list**, which contains other `client-reply content` pairs as well, and wait for better timing to do the write action (next event loop cycle period). In the previous version, Redis read this pending list sequentially, and write the reply content to the client one by one using the main thread. In Redis 6, it changed into writing with multiple I/O threads parallelly.
+
+For the reading action, when a command/message is sent to the Redis server, it's described as a file event, which will trigger Redis to read from the client connection directly. After that, Redis continues to decode/analyze/execute the command/message and get the result, added it to the write pending list as we discussed above, and handles the next happened event until no more available file event or time event happened. So in this case, the `io-threads-do-reads` option allows you to stop reading client connections when the event arrived, collect them to a **pending list** as well, and postpone the read action to the next cycle period using multiple I/O threads. When reads are done, Redis use the main thread to execute the client command one by one.
+
+I/O threads are only activated when the main thread gets the read/write pending list prepared. The workflow can be described as below:
+
+1. main thread prepare the pending list for read
+2. I/O threads read from the connections parallelly
+3. main thread executes command one by one
+4. main thread prepare the pending list for write
+5. I/O threads write reply content to the connections parallelly
+
+
+## Low-Level Design
+### Where is the pending list
```

Comment:
My take is that this section is way too detailed. Someone will probably just go read the code if they want to understand this low level of detail. I think if you remove all of the specific implementation details (names of the arguments, functions, etc), and just leave what the algorithm looks like, it will probably be 50% shorter and just as readable.

I would also say you maybe should fold the high-level/low-level design together and just describe the write flow and then the read flow. Although they have some similarities, and you were trying to generalize, they end up being pretty different.